File: 455.py
class Solution:
    # def findContentChildren(self, g, s):
    def findContentChildren(self, g: List[int], s: List[int]) -> int:
        g.sort()
        s.sort()
        count = 0
        start_j = 0

        for i in g:
            if start_j == len(s):
                return count

            for j, size in enumerate(s[start_j:]):
                if size >= i:
                    count += 1
                    start_j += j+1
                    break
                if j == len(s) - 1:
                    return count

        return count


class Solution:
    # optimize
    def findContentChildren(self, g: List[int], s: List[int]) -> int:
        g.sort()
        s.sort()
        count = 0
        m, n = len(g), len(s)
        i, j = 0, 0

        while i < m and j < n:
            while j < n and s[j] < g[i]:
                j += 1
            if j < n:
                count += 1

            i += 1
            j += 1

        return count


s = Solution()
print(s.findContentChildren([10, 9, 8, 7, 10, 9, 8, 7], [10, 9, 8, 7]))

File: leetcode/花花酱题集/455.js
/**
 * @param {number[]} g
 * @param {number[]} s
 * @return {number}
 */
var findContentChildren = function (g, s) {
  g.sort((x, y) => x - y);
  s.sort((x, y) => x - y);
  let res = 0;
  let i = 0;
  let j = 0;

  while (i < g.length && j < s.length) {
    if (g[i] <= s[j]) {
      i++;
      res++;
    }
    j++;
  }
  return res;
};

g = [10, 9, 8, 7];
s = [5, 6, 7, 8];

console.log(findContentChildren(g, s));

7, 8, 9, 10;
5, 6, 7, 8;
